Pendula95 webclient o auth2 provider

pmlopes opened the pull request:
Motivation:

Prepare merge of #1999 #2000

pmlopes commented:
@pendula95 I've fixed the codegen issue in auth and added replaced the json in your PR with `Credentials` WDYT?

My take here is that if a user uses `Credentials` then they can use:

```java
new UsernamePasswordCredetials("user", "pass");
new TokenCredentials("header.body.signature");
new OAuth2Credentials(...);
```

And it will always be safe, no typpos, etc...

pendula95 commented:
> @pendula95 I've fixed the codegen issue in auth and added replaced the json in your PR with `Credentials` WDYT?
> 
> My take here is that if a user uses `Credentials` then they can use:
> 
> ```java
> new UsernamePasswordCredetials("user", "pass");
> new TokenCredentials("header.body.signature");
> new OAuth2Credentials(...);
> ```
> 
> And it will always be safe, no typpos, etc...

This change LGTM. 
I only have 2 major concerns for previouse comments that I think we should look in to:

- [ ] Explicitly limit only to few oAuth2 flows, Client Credentials & Password Credentials
- [ ] If token expires in between the check and actual request

Rest of the issues I mention can be added later if need and are not mandatory for initial functionality

pmlopes commented:
@pendula95 good points.


###  Explicitly limit only to few oAuth2 flows, Client Credentials & Password Credentials

To address the allowed flows, we should fix this in vertx-auth-oauth2. Currently the module expects a single "flow" which isn't correct. If we look to discovery responses (for example: https://accounts.google.com/.well-known/openid-configuration ) we can see:


grant_types_supported |  
-- | --
0 | "authorization_code"
1 | "refresh_token"
2 | "urn:ietf:params:oauth:grant-type:device_code"
3 | "urn:ietf:params:oauth:grant-type:jwt-bearer"

This means providers can handle more than 1 flow. So, instead of forcing a single flow in `vertx-auth-oauth2` we should make this a list and if discovery is used, we should populate it automatically.

Given that we now know the flows the IdP server is capable, we can simply block requests that are not for those flows.

###  If token expires in between the check and actual request

The oauth2 module works with leeway regarding time validations. By default it's `0` but we can document that for users who need to be able to refresh tokens earlier, we can also had a "about to expire" boolean check that uses a leeway to compute if we should go ahead and refresh or wait for the next time.

The complex scenario here is that how do we define "about to expire"?

pendula95 commented:
Ok, so 

> Explicitly limit only to few oAuth2 flows, Client Credentials & Password Credentials

Will be resolved by discovery mechanism on vertx-auth-oauth2

> If token expires in between the check and actual request

Yes the leeway can fix this issues and I propose to add it by default in the WebClientSession oAuth2Auth that is passed just to cover the marginal case. As the JWT is maintained by the Session is should guarantee that we always use a valid token. This should be an easy fix on this line https://github.com/vert-x3/vertx-web/blob/66d960fc6c41dcea265c360f8618edfb0667729e/vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/SessionAwareInterceptor.java#L78
Just use method `default boolean expired(int leeway)` instead with a default value of 5 seconds.

On the other hand there is an issue when the token is invalidate by other means, example oauth2 provider. In the current implementation I am using an offline check for validity and expiration `User.expired()` and this is faulty as the token can be invalidate and the user request would hit an error that could be hard to recover from. 
This can be fixed by implementing a fail handler on the original request or by using token Token Introspection to check the validity online with oauth2 provider but as mention in the documentation this is not supported by all oauth2 providers so in my opinion the fail handler on the original request would be an better option but I am not sure how to execute this in the limited scope of  `SessionAwareInterceptor`

pmlopes commented:
@pendula95 and @vietj ,

as discussed previously, we should ensure that the webclient keeps a recognizable API with any added extension, plus all extra dependencies should be optional to keep the final artifact small.

As we want to get this merged for 4.2.0 I've reworked the original PR to match these requirements.

In the unit tests you can see that you can create a `WebClientOAuth2` aware by wrapping an existing client. Be it a plain `WebClient` or already a wrapped one e.g.: `WebClientSession` aware.

Please give it a review and we agree with this style, we can finish this quickly by adding a manual entry and examples (plus all the missing javadocs + license headers)

pendula95 commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/WebClientOauth2Aware.java:
```diff
@@ -0,0 +1,78 @@
+  private Credentials credentials;
+
+  private User user;
+  private boolean withAuthentication;
```
This field should be removed as now we don't need explicit call to know if we are going to apply the wrapper logic. We can check instead credentials field.

pmlopes commented:
@pendula95 thanks for the review. I've removed the superfluous field and added tests to verify all the cases in the interceptor + the leeway support.

I think if @vietj agrees, from now on, it's just documentation fixes, and we can merge this for the next beta!

pendula95 commented:
This is the only concern I have #2041

pendula95 commented on vertx-web-client/src/test/java/io/vertx/ext/web/client/WebClientSessionOauth2Test.java:
```diff
@@ -492,4 +501,58 @@ public void testWithAuthenticationWithoutSessionExpiredWithLeeway() throws Excep
+    server = vertx.createHttpServer().requestHandler(req -> {
+      if (req.method() == HttpMethod.POST && "/oauth/token".equals(req.path()) && !retry.get()) {
+        assertEquals("Basic Y2xpZW50LWlkOmNsaWVudC1zZWNyZXQ=", req.getHeader("Authorization"));
+        req.response().putHeader("Content-Type", "application/json").end(loggedOutFixture.encode());
```
forgot to add retry.set(true)

pmlopes replied:
Yes it was a bit strange so I flipped the flag after sending the 401, but it makes more sense here.

pmlopes commented on vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/OAuth2AwareInterceptor.java:
```diff
@@ -36,29 +40,42 @@ public void handle(HttpContext<?> context) {
   }
 
-  private void processFailure(HttpContext<?> context) {
+  private void processResponse(HttpContext<?> context) {
```
this whole section should be guarded by a `enable/disable` flag for users who wish to have full control and not rely on automatic retries.

pendula95 commented:
LGTM!

pmlopes commented:
Thanks to everyone involved! :+1:
